python/cuspatial/cuspatial/core/_column/geocolumn.py

```diff
@@ -30,23 +31,118 @@ def __init__(
         data: Tuple,
         meta: GeoMeta = None,
         shuffle_order: cudf.Index = None,
+        from_read_polygon_shapefile=False,
     ):
-        if (
-            not isinstance(data[0], cudf.Series)
-            or not isinstance(data[1], cudf.Series)
-            or not isinstance(data[2], cudf.Series)
-            or not isinstance(data[3], cudf.Series)
+        if from_read_polygon_shapefile:
+            """
+            A cudf.ListSeries needs four levels of nesting to represent
+            a polygon shapefile. The rings_offsets and polygons_offsets
+            have already been computed in the `read_polygon_shapefile`
+            function. In order to convert it into an arrow list<...>
+            we need a set of offsets buffers for each point tuple, and
+            an offsets buffer for each individual point.
+
+            Coordinates: List of length 2 offsets: [0, 2, 4, ... n/2]
+            Points: List of length 1 offsets: No multipoints
```

Comment:
I don't think this level of indirection is needed. Each element of a ring must be a single-point. With this indirection a polygon type becomes `List<List<List<List<double>>>>`, with the last two level grouping points into multipoints.
But geoarrow spec is `List<List<List<double>>>`.

Reply:
I agree. I tried to do `List<List<List<double>>>` but it did not match what was produced by `from_geopandas` on the same dataset, which lead to discovering that `ListColumn` appears to be producing the `coordinates` and `points` buffers,

Reply:
Can you give a reproducer for this example? It sounds to me if `from_geopandas` is doing something funky to introduce an extra level of indirection.

Reply:
We figured out chatting that the extra buffer is for the `Multipolygons` offset array, which for a set of polygons with only exteriors, matches the `polygons` offset array exactly.